datumaro/plugins/synthetic_data/image_generator.py

```diff
@@ -0,0 +1,208 @@
+            self._width,
+        )
+        use_multiprocessing = sys.platform != "darwin" or sys.version_info > (3, 7)
+        if use_multiprocessing:
+            with Pool(processes=self._cpu_count) as pool:
+                params = pool.map(
+                    self._generate_category, [Random(i) for i in range(self._categories)]
+                )
+        else:
+            params = []
+            for i in range(self._categories):
+                param = self._generate_category(Random(i))
+                params.append(param)
+
+        instances_weights = np.repeat(self._weights, self._instances, axis=0)
+        weight_per_img = np.tile(instances_weights, (self._categories, 1))
+        repeated_params = np.repeat(params, self._weights.shape[0] * self._instances, axis=0)
+        repeated_params = repeated_params[: self._count]
+        weight_per_img = weight_per_img[: self._count]
+        assert weight_per_img.shape[0] == len(repeated_params) == self._count
+
+        splits = min(self._cpu_count, self._count)
+        params_per_proc = np.array_split(repeated_params, splits)
+        weights_per_proc = np.array_split(weight_per_img, splits)
+
+        generation_params = []
+        offset = 0
+        for i, (param, w) in enumerate(zip(params_per_proc, weights_per_proc)):
+            indices = list(range(offset, offset + len(param)))
+            offset += len(param)
+            generation_params.append((i, Random(i), param, w, indices))
+
+        if use_multiprocessing:
+            with Pool(processes=self._cpu_count) as pool:
+                pool.starmap(self._generate_image_batch, generation_params)
+        else:
+            for i, param in enumerate(generation_params):
+                self._generate_image_batch(*param)
+
+    def _generate_image_batch(
+        self, job_id: int, rng: Random, params: np.ndarray, weights: np.ndarray, indices: List[int]
+    ) -> None:
+        print(f"{job_id}: started")
+        proto = osp.join(self._path, "colorization_deploy_v2.prototxt")
+        model = osp.join(self._path, "colorization_release_v2.caffemodel")
+        npy = osp.join(self._path, "pts_in_hull.npy")
+        pts_in_hull = np.load(npy).transpose().reshape(2, 313, 1, 1).astype(np.float32)
+
+        content = open_text(__package__, "synthetic_background.txt")
+        synthetic_background = np.loadtxt(content)
+
+        print(f"Job {job_id}: loading caffe model")
+        net = cv.dnn.readNetFromCaffe(proto, model)
+        net.getLayer(net.getLayerId("class8_ab")).blobs = [pts_in_hull]
+        net.getLayer(net.getLayerId("conv8_313_rh")).blobs = [np.full([1, 313], 2.606, np.float32)]
+
+        print(f"Job {job_id}: starting loop")
+        for i, param, w in zip(indices, params, weights):
+            print(f"Job {job_id}: processing image #{i}")
+            image = self._generate_image(
+                rng, param, self._iterations, self._height, self._width, draw_point=False, weight=w
```

Comment:
I can see from how the RNG is used here that the state of the RNG passed to `_generate_image` depends on how the total number of images is split between CPUs. Therefore, the same command may produce different results on different computers if they have different numbers of CPUs. Perhaps it would be better to create the RNG here (as `Random(i)`) to ensure that the generated results are system-independent?